lib/db/util.js

```diff
@@ -7,6 +7,9 @@
 const crypto = require('crypto')
 const P = require('../promise')
 const randomBytes = P.promisify(require('crypto').randomBytes)
+const scryptHash = P.promisify(require('scrypt-hash'))
+
+const RECOVERY_CODE_CHARS = 'abcdefghijklmnopqrstuvwxyz0123456789'
```

Comment:
@jvehent @g-k From last security meeting we mentioned using alphanumeric characters. Is there any objections to using only lower case or uppercase? Possibly removing O/0, D/O?

Reply:
tl;dr we'll get back to you

> Is there any objections to using only lower case or uppercase?

It reduces the alphabet size. But also probably increases usability and we can always adjust the recovery code length to hit a target key space size.

I'd be interested in research on whether it's easier to use longer codes with a limited alphabet or shorter codes with a larger alphabet for a given keyspace size.

Also, the right font could make 0/O and 1/l more distinguishable, but I'm not sure we want to rely on that.

OWASP just suggests "8 or more characters" for 2fa password reset codes  https://www.owasp.org/index.php/Forgot_Password_Cheat_Sheet#Step_3.29_Send_a_Token_Over_a_Side-Channel

Julien is on PTO for the rest of the week, so we'll have to wait to get his opinion. I'll run this by the rest of the team next Wednesday too.

Reply:
@g-k That sounds good. In the meantime, I should be able to make the keyspace and length for recovery codes configurable (assuming only alphanumberic) so the PR doesn't have to block on it.

FWIW, we have been testing with recovery codes that are 8 hex chars long and they are still hard to read on smaller screens, ref: https://github.com/mozilla/fxa-content-server/issues/6057

Reply:
Thanks! Perfect timing for that issue. There's probably a reason a lot sites use 6-8 digits for 2fa codes.

Reply:
@vbudhram I ran this by some more people. It is good to maximize the recovery code alphabet
size, using a single case is OK especially if we make the rate limits on the "use recovery code" endpoint stricter.

The main threat we could think of was an attacker who already has a password trying to brute force a recovery code to bypass 2FA.

To make this harder we require the password first and can adjust:

* recovery code length
* code alphabet size
* rate limits on the "use recovery code" endpoint
* fraud detection to catch and stop brute forcing attempts sooner

:jrgm said the rate limit for the recovery code endpoint is the same as the login limit: 10 attempts in 15 minutes or ~350k attempts per year.

Estimates to brute force an eight-character recovery code with various alphabets at the current rate limits:

| alphabet name                 | combinations for eight char code | years to brute force a code |
|---------------------------|----------------------------------|--------------------------------------------------|
| digits                    | 10 ^ 8                           | 285                                              |
| single case hex           | 16 ^ 8                           | 12271                                            |
| mixed case hex            | 22 ^ 8                           | 156788                                           |
| single case alphanumeric  | 36 ^ 8                           | 8060314                                          |
| mixed case alphanumeric   | 62 ^ 8                           | 623828873                                        |


For the larger alphabets, I made some notes on potentially ambiguous / homoglyph ASCII character sets here: https://gist.github.com/g-k/2d6a57b8153fcd8196c6adc2a8ef4e93

Reply:
Yeah, good point. It should be per email / UID so the `MAX_BAD_LOGINS` env var https://github.com/mozilla/fxa-customs-server/blob/0ca8c5206405981c35428c08f6e475c4ad044152/lib/config/config.js#L64-L69 unless there's a separate one for recovery code attempts.

Reply:
@g-k Thanks for following up!

> Worth clarifying, is this limit per email, per IP address, or per (email, ip) pair?

Custom server states that login is [per (email, ip)](https://github.com/mozilla/fxa-customs-server#policies).

I created an [issue](https://github.com/mozilla/fxa-customs-server/issues/238) in custom server to add a rate limit rule specifically for recovery codes.